ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp

```diff
@@ -0,0 +1,1210 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <algorithm>
+#include <ngraph/opsets/opset7.hpp>
+
+#include "ngraph/runtime/reference/broadcast.hpp"
+#include "ngraph/runtime/reference/matmul.hpp"
+#include "ngraph/runtime/reference/multiply.hpp"
+#include "ngraph/runtime/reference/reshape.hpp"
+#include "ngraph/runtime/reference/sum.hpp"
+#include "ngraph/runtime/reference/transpose.hpp"
+#include "ngraph/shape.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            namespace einsum_details
+            {
+                std::vector<std::pair<size_t, size_t>> compute_einsum_path(size_t num_inputs)
+                {
+                    // TODO: implement algorithm for finding (pseudo-)optimal einsum_path
+                    std::vector<std::pair<size_t, size_t>> einsum_path;
+                    NGRAPH_CHECK(num_inputs > 0);
+                    for (size_t input_ind = num_inputs - 1; input_ind > 0; --input_ind)
+                    {
+                        einsum_path.push_back(std::make_pair(0, input_ind));
+                    }
+                    return einsum_path;
+                }
+
+                bool is_range_0_to_n(const std::vector<int64_t>& labels_inds)
+                {
+                    int64_t check_index = 0;
+                    for (auto index : labels_inds)
+                    {
+                        if (check_index != index)
+                        {
+                            return false;
+                        }
+                        ++check_index;
+                    }
+                    return true;
+                }
+
+                bool is_dimension_reduced(const std::vector<std::string>& input_subscripts,
```

Comment:
This function is similar to is_dimension_reduced() from einsum_decomposition.cpp. Can it be reused here?

Reply:
Absolutely, agree with you. The situation is not trivial here because we have no common utils routine for both transformations and nGraph core. I have already tried this way but failed. The duplication including three function (compute_einsum_path, is_dimension_reduced, etc) is about 100 lines not too much. So let us keep these duplications.
Regarding the way to include `is_dimension_reduced` into nGraph op API I consider that this function is very specific to implementation and it is not likely that plugin will use this way.